File: MethodChain/method_chain.cpp
#include <iostream>
using namespace std;

class Calculator {
private:
    double result;

public:
    Calculator() : result(0) {}

    Calculator& add(double value) {
        result += value;
        return *this;
    }

    Calculator& subtract(double value) {
        result -= value;
        return *this;
    }

    Calculator& multiply(double value) {
        result *= value;
        return *this;
    }

    Calculator& divide(double value) {
        if (value != 0) {
            result /= value;
        }
        return *this;
    }

    double get_result() const {
        return result;
    }
};

int main() {
    Calculator calc;
    double result = calc.add(5)
                        .multiply(2)
                        .subtract(3)
                        .divide(2)
                        .get_result();

    cout << "결과: " << result << endl;  // 결과: 3.5
    return 0;
}

File: MethodChain/method_chain.kt
class Calculator {
    private var result: Double = 0.0

    fun add(value: Double): Calculator {
        result += value
        return this
    }

    fun subtract(value: Double): Calculator {
        result -= value
        return this
    }

    fun multiply(value: Double): Calculator {
        result *= value
        return this
    }

    fun divide(value: Double): Calculator {
        if (value != 0.0) {
            result /= value
        }
        return this
    }

    fun getResult(): Double {
        return result
    }
}

fun main() {
    val calc = Calculator()
    val result = calc.add(5.0)
                     .multiply(2.0)
                     .subtract(3.0)
                     .divide(2.0)
                     .getResult()

    println("결과: $result") // 출력: 결과: 3.5
}
